CommonCompilerConfig.cmake

```diff
@@ -0,0 +1,184 @@
+# Numenta, Inc. a separate commercial license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# -----------------------------------------------------------------------------
+
+# Configures common compiler/linker/loader settings for internal and external
+# sources.
+#
+# NOTE SETTINGS THAT ARE SPECIFIC TO THIS OR THAT MODULE DO NOT BELONG HERE.
+
+# INPUTS:
+#
+# PLATFORM: lowercase ${CMAKE_SYSTEM_NAME}
+
+# OUTPUTS:
+#
+# BITNESS: Platform bitness: 32 or 64
+#
+# COMMON_COMPILER_DEFINITIONS: list of -D define flags for the compilation of
+#                               source files; e.g., for cmake `add_definitions()`
+# COMMON_COMPILER_DEFINITIONS_STR: string variant of COMMON_COMPILER_DEFINITIONS
+#
+# COMMON_C_FLAGS: string of common C flags
+#
+# COMMON_CXX_FLAGS: string of common C++ flags (COMMON_C_FLAGS + C++ flags)
+#
+# COMMON_LINK_FLAGS: string of linker flags for linking executables and shared
+#                    libraries (DLLs)
+#
+# CMAKE_LINKER: updated, if needed; use ld.gold if available. See cmake
+#               documentation
+
+if(NOT DEFINED PLATFORM)
+    message(FATAL_ERROR "PLATFORM property not defined: PLATFORM=${PLATFORM}")
+endif()
+
+string(TOUPPER ${PLATFORM} PLATFORM_UPPERCASE)
+
+# Identify platform "bitness".
+if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+  set(BITNESS 64)
+else()
+  set(BITNESS 32)
+endif()
+
+set(stdlib_cxx "")
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+  set(stdlib_common "")
+  set(stdlib_cxx "-stdlib=libc++")
+elseif(${CMAKE_SYSTEM_NAME} MATCHES "Linux")
+  set(stdlib_common "")
+elseif(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
+  set(stdlib_common "")
+endif()
+
+
+# Compiler `-D*` definitions
+set(COMMON_COMPILER_DEFINITIONS)
+
+if(UNIX) # or UNIX like (i.e. APPLE and CYGWIN)
+  set(COMMON_COMPILER_DEFINITIONS
+      ${COMMON_COMPILER_DEFINITIONS}
+      -DHAVE_UNISTD_H)
+elseif(MSVC OR MSYS OR MINGW)
+  set(COMMON_COMPILER_DEFINITIONS
+      ${COMMON_COMPILER_DEFINITIONS}
+      -DPSAPI_VERSION=1
+      -DWIN32
+      -D_WINDOWS
+      -D_MBCS
+      -D_CRT_SECURE_NO_WARNINGS
+      -DNDEBUG
+      -D_VARIADIC_MAX=10
+      -DNOMINMAX
+      -DCAPNP_LITE=1)
+  if(MSYS OR MINGW)
+    set(COMMON_COMPILER_DEFINITIONS
+        ${COMMON_COMPILER_DEFINITIONS}
+        -DHAVE_UNISTD_H)
+  endif()
+endif()
+
+
+#
+# Set linker (ld)
```

Comment:
I think the linker is needed for the LTO optimizations, which are/can be used under source builds. ld.gold is better/only worked at the time.